pkg/envoy/sds/builder.go

```diff
@@ -0,0 +1,171 @@
+package sds
+
+import (
+	xds_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	xds_auth "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/envoy"
+	"github.com/openservicemesh/osm/pkg/envoy/secrets"
+	"github.com/openservicemesh/osm/pkg/errcode"
+	"github.com/openservicemesh/osm/pkg/identity"
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+type sdsBuilder struct {
+	proxy *envoy.Proxy
+
+	// Service certificate for this proxy
+	proxyCert *certificate.Certificate
+
+	// Requested certs to generate secrets for
+	requestedCerts []string
+
+	trustDomain string
+
+	ListServiceIdentitiesForService func(service.MeshService) []identity.ServiceIdentity
+
+	// SDS Response
+	sdsResources []types.Resource
+}
+
+func newSdsBuilder() *sdsBuilder {
+	return &sdsBuilder{}
+}
+
+func (b *sdsBuilder) SetProxy(proxy *envoy.Proxy) {
+	b.proxy = proxy
+}
+
+func (b *sdsBuilder) SetProxyCert(cert *certificate.Certificate) {
+	b.proxyCert = cert
+}
+
+func (b *sdsBuilder) SetRequestedCerts(requestedCerts []string) {
+	b.requestedCerts = requestedCerts
+}
+
+func (b *sdsBuilder) SetTrustDomain(trustDomain string) {
+	b.trustDomain = trustDomain
+}
+
+func (b *sdsBuilder) SetListServiceIdentitiesForService(listServiceIdentitiesForService func(service.MeshService) []identity.ServiceIdentity) {
+	b.ListServiceIdentitiesForService = listServiceIdentitiesForService
+}
+
+// Build generates SDS Secret Resources based on requested certs in the DiscoveryRequest
+func (b *sdsBuilder) Build() ([]types.Resource, error) {
+	b.getSDSSecrets()
+	return b.sdsResources, nil
+}
+
+func (b *sdsBuilder) getSDSSecrets() {
+	// requestedCerts is expected to be a list of either of the following:
+	// - "service-cert:namespace/service-account"
+	// - "root-cert-for-mtls-outbound:namespace/service"
+	// - "root-cert-for-mtls-inbound:namespace/service-account"
+
+	b.sdsResources = nil
```

Comment:
nit: no need to set the output on the builder, can simply just return this (from the build method as well)